src/compiler/checker.ts

```diff
@@ -28076,15 +28077,59 @@ namespace ts {
         }
 
         function getAssignedClassSymbol(decl: Declaration): Symbol | undefined {
-            const assignmentSymbol = decl && decl.parent &&
-                (isFunctionDeclaration(decl) && getSymbolOfNode(decl) ||
-                 isBinaryExpression(decl.parent) && getSymbolOfNode(decl.parent.left) ||
-                 isVariableDeclaration(decl.parent) && getSymbolOfNode(decl.parent));
-            const prototype = assignmentSymbol && assignmentSymbol.exports && assignmentSymbol.exports.get("prototype" as __String);
-            const init = prototype && prototype.valueDeclaration && getAssignedJSPrototype(prototype.valueDeclaration);
+            const assignmentSymbol = decl && getSymbolOfExpando(decl, /*allowDeclaration*/ true);
+            const prototype = assignmentSymbol?.exports?.get("prototype" as __String);
+            const init = prototype?.valueDeclaration && getAssignedJSPrototype(prototype.valueDeclaration);
             return init ? getSymbolOfNode(init) : undefined;
         }
 
+        function getSymbolOfExpando(node: Node, allowDeclaration?: boolean): Symbol | undefined {

```

Comment:
changes from the move:

1. Return a Symbol instead of a Node, since all callers immediately do that anyway. This is why I moved it into the checker.
2. Add a parameter `allowDeclaration` that allows `node` to be *either* the expando or its declaration. The parameter disables the checks that assert that `node` is the `expando` initialiser.
3. Add a FunctionDeclaration case when `allowDeclaration` is true.